File: backend/src/utils/permissions.ts
import { User, Role, Permission } from '@database/index';

export interface ResolvedPermissions {
  userId: string;
  roles: string[];
  permissions: Set<string>;
}

export async function resolveUserPermissions(userId: string): Promise<ResolvedPermissions> {
  const user = (await User.findByPk(userId, {
    include: [
      {
        model: Role,
        as: 'roles',
        include: [{ model: Permission, as: 'permissions' }],
      },
    ],
  })) as (User & { roles: Array<Role & { permissions: Permission[] }> }) | null;

  if (!user) {
    return { userId, roles: [], permissions: new Set() };
  }

  const roles = user.roles.map((r) => r.slug);
  const permissions = new Set<string>();
  for (const role of user.roles) {
    for (const permission of role.permissions) {
      permissions.add(permission.slug);
    }
  }

  return { userId, roles, permissions };
}

File: backend/app/core/permissions.py
from app.models import Permission, Role, User


def resolve_user_permissions(db, user_id):
    user = db.query(User).filter(User.id == user_id, User.deletedAt.is_(None)).first()
    if not user:
        return {'userId': user_id, 'roles': [], 'permissions': set()}

    role_slugs = []
    permissions = set()
    for role in user.roles:
        role_slugs.append(role.slug)
        perms = db.query(Permission).join(Permission.roles).filter(Role.id == role.id).all()
        for permission in perms:
            permissions.add(permission.slug)

    return {'userId': user_id, 'roles': role_slugs, 'permissions': permissions}